src/main/java/org/opensearch/neuralsearch/search/collector/HybridTopFieldDocSortCollector.java

```diff
@@ -0,0 +1,411 @@
+        return topFieldDocs;
+    }
+
+    public abstract class HybridTopDocSortLeafCollector implements LeafCollector {
+        HybridQueryScorer compoundQueryScorer;
+        boolean collectedAllCompetitiveHits = false;
+
+        @Nullable
+        FieldDoc after;
+        final Sort sort;
+        boolean initializePerSegment;
+
+        public HybridTopDocSortLeafCollector(Sort sort, @Nullable FieldDoc after) {
+            this.sort = sort;
+            this.after = after;
+            this.initializePerSegment = true;
+        }
+
+        @Override
+        public void setScorer(Scorable scorer) throws IOException {
+            if (scorer instanceof HybridQueryScorer) {
+                log.debug("passed scorer is of type HybridQueryScorer, saving it for collecting documents and scores");
+                compoundQueryScorer = (HybridQueryScorer) scorer;
+            } else {
+                compoundQueryScorer = getHybridQueryScorer(scorer);
+                if (Objects.isNull(compoundQueryScorer)) {
+                    log.error(
+                        String.format(Locale.ROOT, "cannot find scorer of type HybridQueryScorer in a hierarchy of scorer %s", scorer)
+                    );
+                }
+            }
+        }
+
+        private HybridQueryScorer getHybridQueryScorer(final Scorable scorer) throws IOException {
+            if (scorer == null) {
+                return null;
+            }
+            if (scorer instanceof HybridQueryScorer) {
+                return (HybridQueryScorer) scorer;
+            }
+            for (Scorable.ChildScorable childScorable : scorer.getChildren()) {
+                HybridQueryScorer hybridQueryScorer = getHybridQueryScorer(childScorable.child);
+                if (Objects.nonNull(hybridQueryScorer)) {
+                    log.debug(
+                        String.format(
+                            Locale.ROOT,
+                            "found hybrid query scorer, it's child of scorer %s",
+                            childScorable.child.getClass().getSimpleName()
+                        )
+                    );
+                    return hybridQueryScorer;
+                }
+            }
+            return null;
+        }
+
+        /*
+        The method initializes once per search request.
+         */
+        void initializePriorityQueuesWithComparators(LeafReaderContext context, int length) throws IOException {
+            if (compoundScores == null) {
+                numberOfSubQueries = length;
+                compoundScores = new FieldValueHitQueue[length];
+                comparators = new LeafFieldComparator[length];
+                queueFull = new boolean[length];
+                collectedHits = new int[length];
+                for (int i = 0; i < length; i++) {
+                    initializeLeafFieldComparators(context, i);
+                }
+            }
+            if (initializePerSegment) {
```

Comment:
when `initializePerSegment` can be false? We're setting it to `true` in constructor

Reply:
Consider there are 2 shards and each shard has 3 segments.

The leaf collector gets initialized per segment. 
Therefore, the code which is under initializePerSegment will run per segment once during initialization.

As soon as that code runs, it will set the variable to false.